### ISSUE TITLE
TOML as an RFC?

### ISSUE THREAD

**Issue by mnot** [toml-lang/toml#870]
Have you considered publishing v1.0.0 as an [RFC](https://rfc-editor.org/)?

I can see two immediate benefits:

* The format would be easier to use / reference from other RFCs and by other standards efforts.
* You could get `application/toml` [registered](https://www.iana.org/assignments/media-types/media-types.xhtml)[^1].

If you'd consider it, I'd be happy to help - you don't have to go through the full standards process, the independent stream would be very appropriate for this and comparatively lightweight. 

[^1]: You don't need an RFC for the full format to get the registration -- but it does require some sort of standards publication.

**Comment by paulehoffman** [comment]
+1 to the "happy to help". @mnot and I both have written dozens of RFCs (although none together, yet).

**Comment by pradyunsg** [comment]
> Have you considered publishing v1.0.0 as an [RFC](https://rfc-editor.org/)?

Yes! That's the motivation behind keeping the language grammer in ANBF.

IIUC, the blocker here is one of the authors figuring out the "paperwork" of the process and doing it. Help would very much be welcome! :)

**Comment by paulehoffman** [comment]
If one of the main contributors wants to step up as the primary author, I'm happy to walk you through the paperwork. @mnot will probably do so as well, and may agree with me on some of it. :-)

**Comment by mnot** [comment]
Definitely. Consider us at your disposal. From our side, we just need buy-in from the TOML project and someone to work with; we can handle the process issues, etc.

**Comment by eksortso** [comment]
I'm just a bit player here. But with a number of serious enhancements in the pipeline, TOML v1.1.0 would be a better candidate for an RFC once it's released. Is there an immediate demand for an RFC to be filed now, or can we take the time to propose one with TOML's new features?

Deferring until the v1.1.0 release would also give us an opportunity to move toward a set of documents that would line up more with the requirements of an RFC, where it would make sense to further that process in the project as it stands.

**Comment by ChristianSi** [comment]
Yes, I agree. The changes made since TOML 1.0 are significant and certainly justify the release of 1.1 in the foreseeable future. So, to be really future-proof, that's a better target for standardization.

**Comment by mcarans** [comment]
Is there a plan for releasing 1.1? I think it would be a good prompt for library implementers to support the significant changes that have been made since 1.0. That in turn will hopefully prompt some developers who've been reticent to use TOML (and specifically pyproject.toml) to reconsider since I think the changes overcome concerns or limitations that they had raised.

**Comment by eksortso** [comment]
@mcarans This would be better discussed in #915, I think. It's a little off-topic here, though I knew v1.1.0 would be released at some point (there's no timeframe set yet) when I posted a few weeks ago.

This is a significant subject, I know, because Python 3.11 was released yesterday, and it has built-in support for TOML parsing in its standard library (the [`tomllib`](https://docs.python.org/3.11/library/tomllib.html) module, based on [`tomli`](https://github.com/hukkin/tomli). Standard libraries change slowly, and I know that. But the exposure to the format will draw interest from new users.

**Comment by mcarans** [comment]
@eksortso Thanks for you reply. I have put my comment where you suggested. I'll just say here for completeness that I see a [PR in tomli](https://github.com/hukkin/tomli/pull/200) to introduce one of the most important changes since TOML 1.0 but it looks like it is stuck in draft awaiting that change to be official (ie. in a 1.1 release).

**Comment by paulehoffman** [comment]
I'm happy to help with the mechanical part of the RFC process, including the Markdown stuff.

**Comment by robertlagrant** [comment]
Just wondering if this is likely to happen. I don't see it in [the registry](https://www.iana.org/assignments/media-types/media-types.xhtml). We have a use for it!

**Comment by eksortso** [comment]
Your use may give this process a boost! What's your use case?

**Comment by aamelnikov** [comment]
Hi,
I am one of designated expert reviewers for Media Types appointed by IETF and I just received a request to register application/toml using the following template:

-----------------
Name: Ben van Hartingsveldt
Email: ben.vanhartingsveldt@yocto.com
Media type name: application
Media subtype name: toml
Required parameters: N/A
Optional parameters: N/A
Encoding considerations: 8bit
Security considerations: (1) TOML files don't contain executable content.
(2) Information contained in TOML files do not require privacy or integrity services. If a client is storing information that is potentially sensitive, some sort of security is recommended. TLS, and specifically HTTPS, are the recommended way to protect potentially sensitive information.

Interoperability considerations: N/A
Published specification: https://toml.io/
Applications which use this media: Various
Fragment identifier considerations: N/A
Restrictions on usage: N/A
Provisional registration? (standards tree only): No
Additional information:

1. Deprecated alias names for this type: none
2. Magic number(s): none
3. File extension(s): .toml
4. Macintosh file type code: TEXT
5. Object Identifiers: none

General Comments:

Person to contact for further information:

1. Name: Ben van Hartingsveldt
2. Email: ben.vanhartingsveldt@yocto.com

Intended usage: COMMON

Author/Change controller: Ben van Hartingsveldt
-----------------

Do you have any objections/support/have correction(s) to this registration?

In particular note that the "Change controller" is the person who is contacted when further changes (for example if you decide to publish an RFC describing TOML format) to the media type registration are suggested. (Also note that there is no implication that the Change Controller of the registration has any control over the underlying format.)

Best Regards,
Alexey

**Comment by paulehoffman** [comment]
The TOML project seems to be either dead (stuck at 1.0.0) or badly wedged; see https://github.com/toml-lang/toml/issues/928

Having a media type that does not specify the version may be acceptable, but is possibly dangerous, given that the proposed changes in 1.1.0 are breaking changes to the allow bytes in the format (that is, new restrictions on which Unicode charaters are allowed). OTOH, I'm sure that there are similar problems with other formats that have media types.

**Comment by arp242** [comment]
> In particular note that the "Change controller" is the person who is contacted when further changes (for example if you decide to publish an RFC describing TOML format) to the media type registration are suggested. (Also note that there is no implication that the Change Controller of the registration has any control over the underlying format.)

I don't think this person has had any involvement in TOML (AFAIK); I don't think it's a good idea to record them as the contact.

I don't really know who would be; the project is in a "absent maintainer, but not handed over to anyone else"-state.

Also:

> Encoding considerations: 8bit

I'm not sure what "8bit" refers to exactly here, but TOML files are always UTF-8 (which is "8 bit" for some but not all meanings of it).

**Comment by aamelnikov** [comment]
> > In particular note that the "Change controller" is the person who is contacted when further changes (for example if you decide to publish an RFC describing TOML format) to the media type registration are suggested. (Also note that there is no implication that the Change Controller of the registration has any control over the underlying format.)
> 
> I don't think this person has had any involvement in TOML (AFAIK); I don't think it's a good idea to record them as the contact.
> 
> I don't really know who would be; the project is in a "absent maintainer, but not handed over to anyone else"-state.

Thank you.

> Also:
> 
> > Encoding considerations: 8bit
> 
> I'm not sure what "8bit" refers to exactly here, but TOML files are always UTF-8 (which is "8 bit" for some but not all meanings of it).

This field is actually an enum with 3 choices: 7bit, 8bit and binary. 8bit is anything excluding NUL, so UTF-8 without NUL seems consistent with that.

**Comment by aamelnikov** [comment]
> The TOML project seems to be either dead (stuck at 1.0.0) or badly wedged; see #928
> 
> Having a media type that does not specify the version may be acceptable, but is possibly dangerous, given that the proposed changes in 1.1.0 are breaking changes to the allow bytes in the format (that is, new restrictions on which Unicode charaters are allowed). OTOH, I'm sure that there are similar problems with other formats that have media types.

Existence of 2 versions is probably a "Interoperability consideration" for the registration template.

**Comment by notpushkin** [comment]
@aamelnikov Thanks for looking into this!
Would you recommend getting a provisional registration for TOML in meantime?

**Comment by ChristianSi** [comment]
aamelnikov While all the suggested value seem correct and I suppose this is a good-faith submission, having as change controller someone who seems to be unrelated with the TOML project is not a good idea so I too suggest not to accept this submission for now.

Regarding the state of the project: it's currently stalled, because @pradyunsg, the sole remaining maintainer, has largely become inactive but failed to appoint any successors so far. However, there's still hope that that'll change sooner or later and at that point thing might move swiftly again.

As for incompatible versions, there's actually no need to worry about that. All version, from fairly early drafts (I think) up to the latest 1.1 draft, are backwards compatible. As TOML uses semantic version, the first breaking changes might come with 2.0, and that's certainly a long time off, if it ever comes.

**Comment by paulehoffman** [comment]
I agree that having an unrelated person as the change controller is a bad idea. @ChristianSi: are you maybe willing to step up for that role? There is nearly nothing to do but be available for questions in the foreseeable future.

As for the "incompatible versions" part: is there a full list of diffs from 1.0.0 to the intended 1.1.0? I thought that I saw the allowed characters list in parts of 1.1.0 was smaller, meaning that some valid 1.0.0. documents would become invalid. If I'm wrong about that, I'm happy.

**Comment by ChristianSi** [comment]
@paulehoffman:

> @ChristianSi: are you maybe willing to step up for that role? There is nearly nothing to do but be available for questions in the foreseeable future.

Thanks for the suggestion, but while I'm involved with TOML, others such as @eksortso are even more so, and in any case I'm afraid that without a functional maintainer structure such a decision would be arbitrary.

> As for the "incompatible versions" part: is there a full list of diffs from 1.0.0 to the intended 1.1.0? I thought that I saw the allowed characters list in parts of 1.1.0 was smaller, meaning that some valid 1.0.0. documents would become invalid. If I'm wrong about that, I'm happy.

Yes, see the "unreleased" section at the start of the [Changelog](https://github.com/toml-lang/toml/blob/main/CHANGELOG.md).

There it says: "Relax comment parsing; most control characters are again permitted." So it's the opposite of what you thought: TOML 1.0 is fairly restrictive in what it allows in comments, possibly allowing parsing to fail because of stuff that's commented out, which is unintuitive and might need to nasty surprises, since people generally expect text in comments to be *ignored* rather than cause parsing errors. So TOML 1.1 attempts to correct that as far as reasonably possible, accepting more documents but never rejecting any that were already valid.

**Comment by mnot** [comment]
The change controller doesn't need to be a person -- e.g., if you have a community mailing list that could do it - or maybe even the GitHub issues list (although that might raise some eyebrows?).

Another approach would be to nominate [Tom Preston-Werner](https://tom.preston-werner.com/).

**Comment by paulehoffman** [comment]
Or, maybe leave it blank. It is a "general comment", and I think a correct answer is "None". I like @mnot's suggestion, but that is probably a worse can-o-worms than "None"

**Comment by paulehoffman** [comment]
> > As for the "incompatible versions" part: is there a full list of diffs from 1.0.0 to the intended 1.1.0? I thought that I saw the allowed characters list in parts of 1.1.0 was smaller, meaning that some valid 1.0.0. documents would become invalid. If I'm wrong about that, I'm happy.
> 
> Yes, see the "unreleased" section at the start of the [Changelog](https://github.com/toml-lang/toml/blob/main/CHANGELOG.md).
> 
> There it says: "Relax comment parsing; most control characters are again permitted." So it's the opposite of what you thought: TOML 1.0 is fairly restrictive in what it allows in comments, possibly causing parsing to fail because of stuff that's commented out, which is unintuitive and might lead to nasty surprises, since people generally expect text in comments to be _ignored_ rather than cause parsing errors. So TOML 1.1 attempts to correct that as far as reasonably possible, accepting more documents but never rejecting any that were already valid.

Thanks! I had misread this in some other comment in an issue or PR. I'm happy that alll of what is proposed is forwards-compatible.

It would be grand if @pradyunsg would cede control to one of the many people here who seem eager, we get 1.1 out, and then put it in the freezer.

**Comment by aamelnikov** [comment]
> The change controller doesn't need to be a person -- e.g., if you have a community mailing list that could do it - or maybe even the GitHub issues list (although that might raise some eyebrows?).
> 
> Another approach would be to nominate [Tom Preston-Werner](https://tom.preston-werner.com/).

Tom Preston-Werner should probably be listed as "Author", but doesn't need to be "Change Controller".

**Comment by aamelnikov** [comment]
> @aamelnikov Thanks for looking into this! Would you recommend getting a provisional registration for TOML in meantime?

I had to research this :-). Provisional might be the next best thing, but I think this can be registered in the main registry after some changes to the submitted registration template are made.

**Comment by ben221199** [comment]
> Hi, I am one of designated expert reviewers for Media Types appointed by IETF and I just received a request to register application/toml using the following template:
> 
> ---
> 
> Name: Ben van Hartingsveldt
> Email: [ben.vanhartingsveldt@yocto.com](mailto:ben.vanhartingsveldt@yocto.com)
> Media type name: application
> Media subtype name: toml
> Required parameters: N/A
> Optional parameters: N/A
> Encoding considerations: 8bit
> Security considerations: (1) TOML files don't contain executable content.
> (2) Information contained in TOML files do not require privacy or integrity services. If a client is storing information that is potentially sensitive, some sort of security is recommended. TLS, and specifically HTTPS, are the recommended way to protect potentially sensitive information.
> 
> Interoperability considerations: N/A
> Published specification: https://toml.io/
> Applications which use this media: Various
> Fragment identifier considerations: N/A
> Restrictions on usage: N/A
> Provisional registration? (standards tree only): No
> Additional information:
> 
> 1. Deprecated alias names for this type: none
> 2. Magic number(s): none
> 3. File extension(s): .toml
> 4. Macintosh file type code: TEXT
> 5. Object Identifiers: none
> 
> General Comments:
> 
> Person to contact for further information:
> 
> 1. Name: Ben van Hartingsveldt
> 2. Email: [ben.vanhartingsveldt@yocto.com](mailto:ben.vanhartingsveldt@yocto.com)
> 
> Intended usage: COMMON
> 
> Author/Change controller: Ben van Hartingsveldt
> 
> Do you have any objections/support/have correction(s) to this registration?
> 
> In particular note that the "Change controller" is the person who is contacted when further changes (for example if you decide to publish an RFC describing TOML format) to the media type registration are suggested. (Also note that there is no implication that the Change Controller of the registration has any control over the underlying format.)
> 
> Best Regards, Alexey

Currently, the reviewer accepted on changing the information to the following, before it was sent to IESG for it to be grandfathered:

```
Author: Tom Preston-Werner <tom@mojombo.com>
Change controller: https://github.com/toml-lang/toml/issues/870
Submitter: Ben van Hartingsveldt <ben.vanhartingsveldt@yocto.com>
```

The submission was done with good-faith. Tom is mentioned as Author, I as Submitter and the Change controller is currently this issue #870. During submission, I just entered my own name have a submission already and get the ball rolling. You don't need an RFC to have a media type registered at IANA, but I think it would be a good idea to have eventually TOML standardized in the RFC way.

**Comment by ben221199** [comment]
Hello all,

I just received an email from IANA that the media type `application/toml` is finally officially registered.

The registry can be seen here: http://www.iana.org/assignments/media-types
The registration can be seen here: https://www.iana.org/assignments/media-types/application/toml

I'm looking forward to have TOML standardized by RFC in the future and I'm willing to help with that too, but for now lets celebrate this milestone.

**Comment by arp242** [comment]
TOML is already standardized – right here. Most implementations follow this reasonably well.

**Comment by patcon** [comment]
Appreciate the effort, @ben221199!

re: https://github.com/toml-lang/toml/issues/870#issuecomment-2334291428

> I don't think this person has had any involvement in TOML (AFAIK); I don't think it's a good idea to record them as the contact. -- @arp242 

@aamelnikov can you confirm or deny whether having someone minimally involved in the project is how we should leave it? Regardless, now that it's locked in as a milestone, who do you recommend be listed here? (It seems to me it should reflect someone active on the project, yes?)

**Comment by aamelnikov** [comment]
@patcon Sorry I missed your earlier comment. I think the current registration is fine, as the "Change controller" points to this github issue. I think this gives IANA enough context if somebody contacts IANA and requests changes to the registration.

**Comment by pradyunsg** [comment]
@aamelnikov @ben221199 I'd be happy for the point of contact for the project to be updated to `Pradyun Gedam <mail@pradyunsg.me>`. If there's a particular process for that, please let me know.
